homeassistant/components/openai_conversation/__init__.py

```diff
@@ -27,36 +38,82 @@
     DEFAULT_PROMPT,
     DEFAULT_TEMPERATURE,
     DEFAULT_TOP_P,
+    DOMAIN,
 )
 
 _LOGGER = logging.getLogger(__name__)
+SERVICE_GENERATE_IMAGE = "generate_image"
 
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up OpenAI Conversation from a config entry."""
-    openai.api_key = entry.data[CONF_API_KEY]
-
     try:
         await hass.async_add_executor_job(
-            partial(openai.Engine.list, request_timeout=10)
+            partial(
+                openai.Engine.list,
+                api_key=entry.data[CONF_API_KEY],
+                request_timeout=10,
+            )
         )
     except error.AuthenticationError as err:
         _LOGGER.error("Invalid API key: %s", err)
         return False
     except error.OpenAIError as err:
         raise ConfigEntryNotReady(err) from err
 
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = entry.data[CONF_API_KEY]
+
+    _register_services(hass)
```

Comment:
Since the service is not removed when unloading the config entry, move this to `async_setup` instead.